File: SourceCode/Java/Sorting/Bubblesort.java
void sorttime(int[] B) {
    int i;
    long totaltime;
    int runs;
    double avgtime;

    System.out.println("Doing timings for an array of size " + B.length + " on the basis of " + numruns + " runs");

    swaps = 0;
    compares = 0;
    totaltime = 0;
    for (runs=0; runs<numruns; runs++) {
        for (i=0; i<B.length; i++) {
            A[i] = B[i];
        }
        time1 = System.nanoTime();
        bubblesortinstrument(A);
        time2 = System.nanoTime();
        checkorder(A);
        totaltime += (time2-time1);
    }
    System.out.println("Total time is: " + totaltime + ", numruns is: " + numruns);
    avgtime = (((double)totaltime)/numruns) / 1000000.0;
    System.out.println("Instrumented Standard Bubble Sort: Size " + A.length + ", Time: " + avgtime);
    System.out.println("Compares: " + compares/numruns + ", swaps: " + swaps/numruns);
    
    totaltime = 0;
    for (runs=0; runs<numruns; runs++) {
        for (i=0; i<B.length; i++) {
            A[i] = B[i];
        }
        time1 = System.nanoTime();
        bubblesort(A);
        time2 = System.nanoTime();
        checkorder(A);
        totaltime += (time2-time1);
    }
    System.out.println("Total time is: " + totaltime + ", numruns is: " + numruns);
    avgtime = (((double)totaltime)/numruns) / 1000000.0;
    System.out.println("Standard Bubble Sort: Size " + A.length + ", Time: " + avgtime);

    totaltime = 0;
    for (runs=0; runs<numruns; runs++) {
        for(i=0; i<B.length; i++) {
            A[i] = B[i];
        }
        time1 = System.nanoTime();
        bubblesortcheck(A);
        time2 = System.nanoTime();
        checkorder(A);
        totaltime += (time2-time1);
    }
    System.out.println("Total time is: " + totaltime + ", numruns is: " + numruns);
    avgtime = (((double)totaltime)/numruns) / 1000000.0;
    System.out.println("Swap Check Bubble Sort: Size " + A.length + ", Time: " + avgtime);

    totaltime = 0;
    for (runs=0; runs<numruns; runs++) {
        for (i=0; i<B.length; i++) {
            A[i] = B[i];
        }
        time1 = System.nanoTime();
        bubblecheckswap(A);
        time2 = System.nanoTime();
        checkorder(A);
        totaltime += (time2-time1);
    }
    System.out.println("Total time is: " + totaltime + ", numruns is: " + numruns);
    avgtime = (((double)totaltime)/numruns) / 1000000.0;
    System.out.println("Wikipedia Bubble Sort: Size " + A.length + ", Time: " + avgtime);
}

// A flag check if a pass did not have any swaps, which lets us quit
void bubblesortcheck(int[] A) {
    for (int i=0; i<A.length-1; i++) {// Insert i'th record
        boolean swaps = false;
        for (int j=1; j<A.length-i; j++)
            if (A[j-1] > A[j]) {
                swap(A, j-1, j);
                swaps = true;
            }
        if(!swaps) {  // Can quit early
            // System.out.println("Quit at " + i);
            break;
        }
    }
}

// Modify the flag to check position of last swap taken
void bubblesortcheck2(int[] A) {
    for (int i=0; i<A.length-1; i++) {// Insert i'th record
        int lastseen = 0;
        int top = A.length;
        for (int j=1; j<top; j++)
            if (A[j-1] > A[j]) {
                swap(A, j-1, j);
                lastseen = j-1;
            }
        top = lastseen;
        if (top == 0) {  // Can quit early
            // System.out.println("Quit at " + i);
            break;
        }
    }
}

// Wikipedia article "optimization" to only swap up to the last swap seen
/* *** ODSATag: BubblesortCheck *** */
void bubblecheckswap(int[] A) {
    int n = A.length;
    while (n > 0) {
        int newn = 0;
        for (int i = 1; i < n; i++) {
            /* if this pair is out of order */
            if (A[i - 1] > A[i]) {
                swap(A, i - 1, i);
                newn = i;
            }
        }
        n = newn;
    }
}
/* *** ODSAendTag: BubblesortCheck *** */

// Wikipedia article "optimization" to only swap up to the last swap seen
void unwikipedia(int[] A) {
    int n = A.length-1;
    while (n>0) {
        for (int i=0; i<n; i++)
            /* if this pair is out of order */
            if (A[i] > A[i+1]) {
                swap(A, i, i+1);
            }
        n -= 1;
    }
}

// Wikipedia article "optimization" rewritten with a for loop
void wikipedia2(int[] A) {
    int newn;
    int loopcnt = 0;
    for(int n=A.length-1; n>0; n=newn) {
        loopcnt++;
        newn = 0;
        for (int i=0; i<n; i++)
            /* if this pair is out of order */
            if (A[i] > A[i+1]) {
                swap(A, i, i+1);
                newn = i;
            }
    }
    // System.out.println("Loopcnt was " + loopcnt);
}


Boolean sorttest(int[] B) {
    int i;
    for (i=0; i<B.length; i++) {
        A[i] = B[i];
    }
    bubblesort(A);
    if (!checkorder(A)) {
        return false;
    }
    return true;
}

/* *** ODSATag: Bubblesort *** */
void bubblesort(int[] A) {
  for (int i=0; i<A.length-1; i++) // Insert i'th record
    for (int j=1; j<A.length-i; j++)
      if (A[j-1] > A[j])
        swap(A, j-1, j);
}
/* *** ODSAendTag: Bubblesort *** */

void bubblesortinstrument(int[] A) {
    for (int i=0; i<A.length-1; i++) // Insert i'th record
        for (int j=1; j<A.length-i; j++) {
            if (A[j-1] > A[j]) {
                swap(A, j-1, j);
                swaps++;
            }
            compares++;
        }
}

File: SourceCode/C++/Sorting/Bubblesort.cpp
#include "utils.h"
#include "Comparable.cpp"
#include "KVPair.cpp"
#include "Int.cpp"
#include "Checkorder.cpp"
//Written by Hannah Börjesson 2014 @LiU

// With KVPair
/* *** ODSATag: Bubblesort *** */
void bubblesort(Comparable* A[], int n) {
  for (int i = 0; i < n-1; i++) // Insert i'th record
    for (int j = 0; j < n-i; j++) 
      if (*A[j] > *A[j+1]) 
        swap(A, j, j+1);
}
/* *** ODSAendTag: Bubblesort *** */

// With KVPair
/* *** ODSATag: BubblesortCheck *** */
void bubblecheckswap(Comparable* A[], int n) {
  n = n-1;
  while (n > 0) {
    int newn = 0;
    for (int i = 0; i < n; i++) {
      /* if this pair is out of order */
      if (*A[i] > *A[i+1]) {
        swap(A, i, i+1);
        newn = i;
      }
    }
    n = newn;
  }
}
/* *** ODSAendTag: BubblesortCheck *** */

bool sorttest(int array[], int n, int threshold) {
  Comparable* A[n];
  int i;

  /* Sort an array of Ints */
  for (i = 0; i < n; ++i) {
    A[i] = new Int(array[i]);
  }

  //  for (i = 0; i < n; ++i) {
  //    cout << *A[i] << " ";
  //  }
  //  cout << std::endl;
  
  bubblesort(A, n);

  if (!checkorder(A, n)) return false;

  for (i = 0; i < n; ++i) {
    delete A[i];
  }

  /* Sort an array of KVPairs */
  
  for (i = 0; i < n; ++i) {
    A[i] = new KVPair(array[i], &array[i]);
  }

  bubblesort(A, n);

  if (!checkorder(A, n)) return false;

  for (i = 0; i < n; ++i) {
    delete A[i];
  }
  
  delete[] array;

  return true;
}


bool sorttime(int array[], int arraysize, int listsize, int threshold) {
  Comparable* A[arraysize];
  int i;

  /* Sort an array of Ints */
  for (i = 0; i < arraysize; ++i) {
    A[i] = new Int(array[i]);
  }

  Settime();
  for (i=0; i<arraysize; i+=listsize) {
    bubblesort(&A[i], listsize);
  }
  cout << "Sort Int array with list size " << listsize
       << ", array size " << arraysize
       << ", and threshold " << threshold << ": "
       << Gettime() << " seconds\n";

  for (i=0; i<arraysize; i+=listsize) {
    if (!checkorder(&A[i], listsize)) return false;
  }

  for (i = 0; i < arraysize; ++i) {
    delete A[i];
  }


  /* Sort an array of KVPairs */
  
  for (i = 0; i < arraysize; ++i) {
    A[i] = new KVPair(array[i], &array[i]);
  }

  Settime();
  for (i=0; i<arraysize; i+=listsize) {
    bubblesort(&A[i], listsize);
  }
  cout << "Sort KVPair with list size " << listsize
       << ", array size " << arraysize
       << ", and threshold " << threshold << ": "
       << Gettime() << " seconds\n";

  for (i=0; i<arraysize; i+=listsize) {
    if (!checkorder(&A[i], listsize)) return false;
  }

  for (i = 0; i < arraysize; ++i) {
    delete A[i];
  }
  

  /* Sort an array of Ints */
  for (i = 0; i < arraysize; ++i) {
    A[i] = new Int(array[i]);
  }

  Settime();
  for (i=0; i<arraysize; i+=listsize) {
    bubblecheckswap(&A[i], listsize);
  }
  cout << "Sort Int array on checking swaps with list size " << listsize
       << ", array size " << arraysize
       << ", and threshold " << threshold << ": "
       << Gettime() << " seconds\n";

  for (i=0; i<arraysize; i+=listsize) {
    if (!checkorder(&A[i], listsize)) return false;
  }

  for (i = 0; i < arraysize; ++i) {
    delete A[i];
  }


  /* Sort an array of KVPairs */
  
  for (i = 0; i < arraysize; ++i) {
    A[i] = new KVPair(array[i], &array[i]);
  }

  Settime();
  for (i=0; i<arraysize; i+=listsize) {
    bubblecheckswap(&A[i], listsize);
  }
  cout << "Sort KVPair on checking swaps with list size " << listsize
       << ", array size " << arraysize
       << ", and threshold " << threshold << ": "
       << Gettime() << " seconds\n";

  for (i=0; i<arraysize; i+=listsize) {
    if (!checkorder(&A[i], listsize)) return false;
  }

  for (i = 0; i < arraysize; ++i) {
    delete A[i];
  }


  return true;
}

#include "SortTest.cpp"
